### PR TITLE
Add dedicated docstrings to Sum/Product impl of Result

### PR DESCRIPTION
**Pull request by birkenfeld**
(and fix a minor grammar typo below)

### CODE REVIEW

File: src/libcore/iter/traits.rs [hunk 1]
```diff
@@ -798,6 +798,21 @@ impl<I, T, E> Iterator for ResultShunt<I, E>
 impl<T, U, E> Sum<Result<U, E>> for Result<T, E>
     where T: Sum<U>,
 {
+    /// Takes each element in the `Iterator`: if it is an `Err`, no further
+    /// elements are taken, and the `Err` is returned. Should no `Err` occur,
+    /// the sum of all elements is returned.
+    ///
+    /// Here is an example which sums up every integer in a vector,
```
Comment (frewsxcv) [inline src/libcore/iter/traits.rs]: As per https://github.com/rust-lang/rfcs/pull/1574, there should be an `# Examples` line before this line

File: src/libcore/iter/traits.rs [hunk 2]
```diff
@@ -798,6 +798,21 @@ impl<I, T, E> Iterator for ResultShunt<I, E>
...
+    /// rejecting the sum if a negative element is encountered:
+    ///
+    /// ```
+    /// let v = vec![1, 2];
+    /// let res: Result<i32, &'static str> = v.iter().map(|&x: &i32|
+    ///     if x < 0 { Err("Negative element found") }
+    ///     else { Ok(x) }
+    /// ).sum();
+    /// assert!(res == Ok(3));
```
Comment (frewsxcv) [inline src/libcore/iter/traits.rs]: `assert_eq!`

File: src/libcore/iter/traits.rs [hunk 3]
```diff
@@ -798,6 +798,21 @@ impl<I, T, E> Iterator for ResultShunt<I, E>
...
+    /// the sum of all elements is returned.
+    ///
+    /// Here is an example which sums up every integer in a vector,
+    /// rejecting the sum if a negative element is encountered:
+    ///
+    /// ```
+    /// let v = vec![1, 2];
+    /// let res: Result<i32, &'static str> = v.iter().map(|&x: &i32|
+    ///     if x < 0 { Err("Negative element found") }
+    ///     else { Ok(x) }
```
Comment (frewsxcv) [inline src/libcore/iter/traits.rs]: Your change from https://github.com/rust-lang/rust/pull/42569/files could be copied here

**Review by frewsxcv — CHANGES_REQUESTED** [review]

File: src/libcore/iter/traits.rs (hunk 1 above)
Comment (birkenfeld) [inline src/libcore/iter/traits.rs]: Ok.

File: src/libcore/iter/traits.rs (hunk 3 above)
Comment (birkenfeld) [inline src/libcore/iter/traits.rs]: Hm, it's not so easy to apply here. Suggestion?

File: src/libcore/iter/traits.rs (hunk 2 above)
Comment (birkenfeld) [inline src/libcore/iter/traits.rs]: Sure.

File: src/libcore/iter/traits.rs (hunk 3 above)
Comment (frewsxcv) [inline src/libcore/iter/traits.rs]: Ah you're right, nevermind

**Comment by frewsxcv** [comment]
@bors r+ rollup

Thanks!

**Pull request merged**